Disambiguate in precedence order when considering extra attributes (#21696)
gradle/gradle: subprojects/docs/src/docs/userguide/dep-man/04-modeling-features/variant_attributes.adoc

    a. For each requested attribute, if a candidate does not have a value matching the disambiguation rule, it's eliminated from consideration.
    b. If the attribute has a known precedence, Gradle will stop as soon as there is a single candidate remaining.
    c. If the attribute does not have a known precedence, Gradle must consider all attributes.
5. If several candidates still remain, Gradle will start to consider "extra" attributes to disambiguate between multiple candidates. Extra attributes are attributes that were not requested by the consumer but are present on at least one candidate. These extra attributes are considered in precedence order.
    a. If the attribute has a known precedence, Gradle will stop as soon as there is a single candidate remaining.
    b. After all extra attributes with precedence are considered, the remaining candidates can be chosen if they are compatible with all of the non-ordered disambiguation rules.
6. If several candidates still remain, Gradle will consider extra attributes again. A candidate can be chosen if it has the fewest number of extra attributes.

> If at any step no candidates remain compatible, resolution will fail and list all compatible candidates from step 1.

Reviewer: ```suggestion
If at any step no candidates remain compatible, resolution fails.
Additionally, Gradle outputs a list of all compatible candidates from step 1.
```

Suggestion: especially when we're talking about a consequence as dire as failure, it's best to use active voice. You could also consider adding a "for ..." to the second sentence here -- I *suspect* this is for debugging, but it might be helpful for readers to know why we output a list. Also switched the second sentence to the verb "outputs" to make it more obvious what Gradle actually does in this situation.
